packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,211 @@
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const operatorPattern = /(\|\||&&)/g;
+const identifierPattern = /^(!?[a-z0-9_]+)$/;
+
+const tokenize = (expression: string) => expression.split(operatorPattern);
+const matchIdentifier = (identifier: string) => identifier.match(identifierPattern);
+
+export const evaluate = (expression: string, env: Env) => {
+  const tokens = tokenize(expression);
+
+  let result = false;
+  let isContinuedByAndOperator = false;
+
+  for (let i = 0; i < tokens.length; i++) {
+    if (i % 2) {
+      if (tokens[i][0] === '|') {
+        isContinuedByAndOperator = false;
+      } else if (tokens[i][0] === '&') {
+        isContinuedByAndOperator = true;
+      } else {
+        // Invalid expression
+        return false;
+      }
+    } else {
+      const match = matchIdentifier(tokens[i]);
+
+      if (!match) {
+        // Invalid expression
+        return false;
+      }
+
+      let identifier = match[1];
+
+      const isNegated = identifier.charCodeAt(0) === 33; /* '!' */
+      let isPositive = false;
+
+      if (isNegated) {
+        identifier = identifier.slice(1);
+        isPositive = !env.get(identifier);
+      } else {
+        isPositive = !!env.get(identifier);
+      }
+
+      if (isContinuedByAndOperator) {
+        result &&= isPositive;
+      } else {
+        result ||= isPositive;
+      }
+    }
+  }
+
+  return result;
+};
+
+export default class Preprocessor {
+  public static parse(line: string): Preprocessor | null {
+    return new this({
+      condition: line.slice(5 /* '!#if '.length */).replace(/ /g, ''),
+    });
+  }
+
+  public static deserialize(view: StaticDataView): Preprocessor {
+    const condition = view.getUTF8();
+
+    const positives = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      positives.add(view.getUint32());
+    }
+
+    const negatives = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      negatives.add(view.getUint32());
+    }
+
+    return new this({
+      condition,
+      positives,
+      negatives,
+    });
+  }
+
+  public readonly condition: string;
+
+  public readonly positives: Set<number>;
+  public readonly negatives: Set<number>;
+
+  public result: boolean | undefined;
+
+  constructor({
+    condition,
+    positives = new Set(),
+    negatives = new Set(),
+  }: {
+    condition: string;
+    positives?: Set<number>;
+    negatives?: Set<number>;
+  }) {
+    this.condition = condition;
+
+    this.positives = positives;
+    this.negatives = negatives;
+  }
+
+  public evaluate(env: Env) {
+    if (!this.result) {
+      this.result = evaluate(this.condition, env);
+    }
```

Comment:
I think it's safe to cache them if we're going to leave this as an internal implementation detail of the FiltersEngine. I aware about it. I'll remove that caching if you think it's unsafe in other situations.

Reply:
Do we save anything if the only time the condition is evaluated is when the condition change anyway? It's not clear if this cached value will ever be used.

Reply:
The value here is expected to be used multiple times by the `build` property of the `PreprocessorBucket` but I've optimised the `build` function to avoid calling preprocessor multiple times. Now the cached value isn't used anymore after the generation unless the user flushes the cache.

The code related to the evaluation cache of preprocessor now removed.